### ISSUE TITLE
[Question] Inspecting the effects currently in `r`

### ISSUE THREAD

**Issue by sandoichi** [isovector/polysemy#70]
If I have a `Sem r a`, is there any way that I can inspect `r` to see what effects are in there?

Ideally I would like to be able to dynamically determine which interpreters to run while there are remaining effects, other than `Lift IO`, in my `Sem`

Something like:
```haskell
case inspectEffects sem of
  Effect1':rs -> runEffect1 sem
  Effect2':rs -> runEffect2 sem
```

**Comment by isovector** [comment]
Sorry, I don't understand the use case. `r` is left intentionally polymorphic so that you can be flexible with its interpretation. You can inspect it by _intentionally writing it concretely_, but then it wouldn't be dynamic, and I'm not sure why you would want to do this.

If you can flesh out more of the example I'm sure we can figure out some way to help!

**Comment by sandoichi** [comment]
Hey, so I'm working on a program that is going to be programmatically configuring/installing things in Windows. As such, I have effects that describe certain external programs I need to interact with (Npm, Powershell, etc).

I'm trying to make the program flexible enough to put together a collection of actions to run (using any number of effects) based on the environment's needs (configured at run time). I would like to avoid having to write specific combinations of interpreters for every combination of effect that  I may hit (I have a 1:1 relationship between effects and interpreters in this program, so I will always interpret effect X the same way, so no need to care about what context it is being used in).

I can't make one giant interpret function that calls every interpreter that I have because some of my interpreters require proof of certain conditions in order to run. For example, you have to provide a path to `npm.cmd` in order to interpret my Npm effect. I can ensure that npm gets installed if it isn't there (thus gaining my path to npm.cmd), but I only want to care about this for environments that need npm. So if I could say `if my r contain an Npm effect, make sure npm is installed on the machine and then interpret that effect` it would make it so that I can make a generic interpret function that can figure out what interpreters to run and what data it needs based on what the current `Sem` has inside of it.

So in summary, I will have a dynamically composed `Sem` based on how the current environment is configured, and would like to make a dynamic interpret function that calls the appropriate interpreters in  order to reduce down to `IO` without being forced to run every interpreter every time, even if they aren't needed.

Or maybe this just completely goes against the design of polysemy and isn't possible?

**Comment by isovector** [comment]
I think I see the problem; though I'm not sure what a solution would actually look like. Conceptually it seems murky to me.

The idea is you have a static workflow of actions you potentially want to perform? You want to do lots of things, one of which might involve npm, and in that case you need to know where `npm.cmd` is?

If that's accurate, you _can_ model this with a static effect stack. Consider this:

```haskell
program :: Members '[Random, NPM] r => Sem r ()
program = do
  bool <- random @Bool
  when bool $ do
    doSomethingWithNPM
    doSomethingWithNPM
```

you can write your NPM effect like this:

```haskell
data NPM m a where
  DoSomethingWithNPM :: NPM m ()

runNPM
    :: (Member (Input FilePath) r, Member (Lift IO) r)
    => Sem (NPM ': r) a
    -> Sem r a
runNPM = interpret \case
  DoSomethingWithNPM -> do
    npm <- input
    liftIO $ putStrLn $ "hello from npm: " ++ npm
```

but now we want to ask the user to help us with that `Input FilePath`: We can
introduce a new effect `Initialize` which will only get called once and cache
its value. We transform our `Input` into an `Initialize`:

```haskell
data Initialize v m a where
  Initialize :: Initialize v m v

runInputAsInitialize
    :: Member (Initialize i)
    => Sem (Input i ': r) a
    -> Sem r a
runInputAsInitialize
  = runState Nothing
  . reinterpret \case
      Input -> do
        value <- get
        case value of
          Just v -> pure v
          Nothing -> do
            v <- initialize
            put $ Just v
            pure v
```

and then can give a prompt to the user as an interpretation of `Initialize`:

```haskell
runInitializeForNPM
    :: Member (Lift IO) r
    => Sem (Initialize FilePath ': r) a
    -> Sem r a
runInitializeForNPM = interpret \case
  Initialize -> sendM $ do
    putStrLn "what is the npm value?"
    getLine
```

Finally we can tie it all together:

```haskell
main = runM
     . runRandom
     . runInitializeForNPM
     . runInputAsInitialize @FilePath
     . runNPM
     $ program
```

which will half the time print nothing (from the `False` random bool case), and
half the time print this:

```
> what is the npm value
polysemy
> hello from npm: polysemy
> hello from npm: polysemy
```

**Comment by isovector** [comment]
This initialization thing seems valuable enough to put into the `polysemy-zoo`

**Comment by sandoichi** [comment]
Oh I think I understand the approach here. So instead of saying
`When we run the npm interpreter, we ALWAYS need a filepath passed in, so only interpret NPM effects on environments that actually use it`

You would instead say
`We are always going to run the NPM interpreter because our program might need it, but it will only attempt to obtain the filepath when it actually needs to run an NPM effect`

So then we can use the cache to hold the NPM filepath, and the first time we request it and it isn't in that cache we know we need to ensure it is installed and find its path.

Did I interpret (no pun intended) your example correctly? Also, thanks for that big code writeup!

**Comment by isovector** [comment]
Yup, absolutely, that's the idea! Is this enough to get you unstuck?

**Comment by sandoichi** [comment]
Yep, I hadn't thought of using stateful effects here, but it completely makes sense! (I'm still mentally transitioning from F# to Haskell..) Thanks for the help!
